### PR TITLE
Bug 1820075: Reflect route status to ingress status

### PR DESCRIPTION
**Pull request by Miciah**
#### ingress: Rename client to routeClient

* `pkg/route/ingress/ingress.go` (`Controller`): Rename "client" to "routeClient".
(`NewController`, `sync`):
* `pkg/route/ingress/ingress_test.go` (`TestController_stabilizeAfterCreate`, `TestController_sync`): Use the new field name.

#### ingress: Rename fake route clientset

* `pkg/route/ingress/ingress_test.go`: Rename import of the fake clientset for the route API to "routev1fake".
(`TestController_stabilizeAfterCreate`, `TestController_sync`): Rename the fake clientset for the route API from "kc" to "routeClientset".  Rename "actions" to "routeActions".

#### ingress: wantCreates, wantPatches, wantDeletes

* `pkg/route/ingress/ingress_test.go` (`TestController_sync`): Rename `wantCreates` to `wantRouteCreates`, `wantPatches` to `wantRoutePatches`, and `wantDeletes` to `wantRouteDeletes`.

#### Reflect route status to ingress status

* `pkg/cmd/controller/ingress.go` (`RunIngressToRouteController`): Pass ingress client to `ingress.NewController`.
* `pkg/route/ingress/ingress.go` (`Controller`): Add `ingressClient` field.
(`NewController`): Add `ingressClient` parameter.  Use the new parameter to set the `Controller` field.
(`sync`): Reflect the canonical hostname from routes that match the ingress to the ingress's status.
* `pkg/route/ingress/ingress_test.go` (`TestController_stabilizeAfterCreate`): Add fake clientset for the networking API.  Use this fake clientset to initialize the controller.
(`TestController_sync`): Add `wantIngressUpdates` to the test inputs.  Add fake clientset for the networking API.  Use `wantIngressUpdates` in new test cases for ingress status updates.  Use the fake networking clientset to initialize the controller and to verify the new tests cases.

----

Depends on https://github.com/openshift/openshift-apiserver/pull/129 and https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/169.

### CODE REVIEW

File: pkg/route/ingress/ingress.go [hunk 1]
```diff
@@ -435,15 +439,45 @@ func (c *Controller) sync(key queueKey) error {
...
 
 	// purge any previously managed routes
 	for _, route := range old {
-		if err := c.client.Routes(route.Namespace).Delete(context.TODO(), route.Name, metav1.DeleteOptions{}); err != nil && !kerrors.IsNotFound(err) {
+		if err := c.routeClient.Routes(route.Namespace).Delete(context.TODO(), route.Name, metav1.DeleteOptions{}); err != nil && !kerrors.IsNotFound(err) {
+			errs = append(errs, err)
+		}
+	}
+
+	// reflect route status to ingress status
```
Comment (sgreene570) [inline pkg/route/ingress/ingress.go:455]: If I'm following these changes correctly, route status should only be reflected back to the parent ingress object if `len(matches) > 0`, right? If that's true, would it be worth encasing lines 456->483 in an if statement that checks for the length of `matches`?

Comment (Miciah) [reply pkg/route/ingress/ingress.go:455]: I don't think that's right.  What if the ingress is created in such a way that corresponding routes are created and their statuses reflected back to the ingress, and then the ingress is modified in such a way that the routes are removed?  The status should be cleared in that case.  Hm, maybe this needs a test case...

Comment (sgreene570) [reply pkg/route/ingress/ingress.go:455]: Ah, I see,I missed that edge case.

In terms of adding a test case, perhaps one of the `delete route when` test cases could be enhanced to verify that Ingress status conditions are updated properly when corresponding routes are deleted? The current code coverage for the changes in this PR is fantastic, but I think this potential flow should be visible from the test cases (or at least some in-line comments).

**Comment by adambkaplan** [comment]
/hold

Shouldn't this require https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/169 to merge first?
Do we have an e2e test which ensures the ingress-to-route controller successfully updates status?

**Comment by Miciah** [comment]
> Shouldn't this require [openshift/cluster-openshift-controller-manager-operator#169](https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/169) to merge first?

Yeah, that has been the piece blocking this PR.

> Do we have an e2e test which ensures the ingress-to-route controller successfully updates status?

No, just unit tests.

File: pkg/route/ingress/ingress.go (hunk 1 above)
Comment (Miciah) [reply pkg/route/ingress/ingress.go:455]: Done.  Thanks!

File: pkg/route/ingress/ingress_test.go [hunk 2]
```diff
@@ -463,6 +469,7 @@ func TestController_sync(t *testing.T) {
 		wantCreates     []*routev1.Route
 		wantPatches     []clientgotesting.PatchActionImpl
 		wantDeletes     []clientgotesting.DeleteActionImpl
+		wantUpdates     []clientgotesting.UpdateActionImpl
```
[the remark this answers is not in the document]
Comment (Miciah) [reply pkg/route/ingress/ingress_test.go]: Done.  I also renamed `wantCreates` to `wantRouteCreates`, `wantPatches` to `wantRoutePatches`, and `wantDeletes` to `wantRouteDeletes`.

**Comment by Miciah** [comment]
Rebased.

**Comment by Miciah** [comment]
The failure apparently wasn't related to ingress-to-route this time:

    FAIL: github.com/openshift/openshift-controller-manager/pkg/serviceaccounts/controllers TestUpdateNewStyleSecretIPv6 45s

/test unit

**Comment by Miciah** [comment]
`TestUpdateNewStyleSecretIPv6` failed again.  Unit tests all pass locally.  One more try, and if it fails again, I'll open a test PR.  If the test PR fails, I'll open a Bugzilla report.
/test unit

**Comment by Miciah** [comment]
H, this time `TestUpdateOldStyleSecretWithoutKey` failed.

**Comment by Miciah** [comment]
Unit tests passed on #135.
/test unit

**Comment by Miciah** [comment]
And now unit tests are passing here.

**Comment by Miciah** [comment]
openshift/cluster-openshift-controller-manager-operator#169 merged, so
/hold cancel

**Review by adambkaplan — COMMENTED** [review]
@Miciah squash commits, and then I think this looks good. I'd like someone from the network-edge team to LGTM before merge.

/approve

**Comment by Miciah** [comment]
> @Miciah squash commits, and then I think this looks good. I'd like someone from the network-edge team to LGTM before merge.

Done (but I think it hurts readability).  Thanks!  @sgreene570, do you mind giving this another look?

**Comment by sgreene570** [comment]
@Miciah @adambkaplan I reviewed these changes again and have no further comments :grin: 
/lgtm

**Comment by sgreene570** [comment]
@Miciah this PR seems to be consistently failing the `: [sig-network] Ingress API should support creating Ingress API operations [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s] ` test.

**Comment by Miciah** [comment]
> @Miciah this PR seems to be consistently failing the `[sig-network] Ingress API should support creating Ingress API operations [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]` test in the e2e-aws job.

The test creates an ingress, futzes with it a bit, and then tries to patch its status:  
https://github.com/kubernetes/kubernetes/blob/b4ffb455315c/test/e2e/network/ingress.go#L977-L1143
But now the ingress-to-route controller is setting the ingress's status.  The immediate problem is that the test doesn't get before patching, so it has a stale copy, so the update fails.  The bigger problem is that the test assumes that nothing else updates the ingress's status.  Because the test does not set `spec.ingressClassName`, such an expectation is _almost_ reasonable, but the API godoc says, "Implementations of this API should ignore Ingresses without a class specified."  Not _"must"_, just _"should"_.  So strictly speaking, ingress-to-route isn't totally nonconformant here.  This all raises some questions:  
* Is the test's expectation that nothing else updates the ingress's status reasonable?  I'd argue it isn't (see "not _'must'_, just _'should'_").  
* Does the ingress-to-route controller need to start respecting `ingressClassName`?  We'd need (a) to invent a class name, (b) to start setting the default class, (c) to add the class to existing ingresses, and (d) to start respecting the class in the ingress-to-route controller.  It's a good bit of effort (especially migrating existing ingresses) but probably a good long-term goal.  
* Should we skip the test for now?  Updating the status doesn't make us any less conformant as we're already _not_ ignoring the ingress, and I'd argue that updating the status makes us _more_ conformant as long as we _aren't_ ignoring the ingress because the godoc of `status.loadBalancer` is, "LoadBalancer contains the current status of the load-balancer."  So we are nonconformant if we expose the ingress as a route and fail to reflect that to the ingress's status.

Thoughts?

**Comment by Miciah** [comment]
/hold
while we debate API semantics.

**Comment by Miciah** [comment]
Another question:  Why wasn't this test failing before?

**Comment by Miciah** [comment]
I suppose one could make the case that an ingress controller must not remove entries from `status.loadBalancer` that the controller did not itself add, because an ingress controller cannot know that it alone exposes  the ingress.  By that line of reasoning, the test is still broken, but ingress-to-route should be more careful not to clear out entries that it did not add.

**Comment by sgreene570** [comment]
> * Should we skip the test for now?  Updating the status doesn't make us any less conformant as we're already _not_ ignoring the ingress, and I'd argue that updating the status makes us _more_ conformant as long as we _aren't_ ignoring the ingress because the godoc of `status.loadBalancer` is, "LoadBalancer contains the current status of the load-balancer."  So we are nonconformant if we expose the ingress as a route and fail to reflect that to the ingress's status.

Yea, I think in this case it would make sense to temporarily skip the test and revisit it later. To your first point, I would also argue that it is unreasonable for the test to expect _no_ changes to the Ingress status post Ingress creation.

> Does the ingress-to-route controller need to start respecting ingressClassName? 

Most likely yes. I agree this is a non-trivial change, but is still a good long term goal.

> By that line of reasoning, the test is still broken, but ingress-to-route should be more careful not to clear out entries that it did not add.

Does this warrant changes to this PR, or are you just making a general note about how the ingress-to-route controller should work?

**Comment by Miciah** [comment]
> > By that line of reasoning, the test is still broken, but ingress-to-route should be more careful not to clear out entries that it did not add.
> 
> Does this warrant changes to this PR, or are you just making a general note about how the ingress-to-route controller should work going forward?

I went ahead and pushed a change to that effect.

**Comment by sgreene570** [comment]
New changes look good to me. 
@Miciah ping me once the test in question passes or is skipped for `lgtm`

**Comment by sgreene570** [comment]
hm, seconds after I left the above comment the e2e job passed.
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_openshift-controller-manager/130/pull-ci-openshift-openshift-controller-manager-master-e2e-aws/1308798128467480576 passed all tests :thinking: 

@Miciah you were under the impression that this PR should continue to fail on `[sig-network] Ingress API should support creating Ingress API operations [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]`?

**Comment by Miciah** [comment]
> hm, seconds after I left the above comment the e2e job passed.
> https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_openshift-controller-manager/130/pull-ci-openshift-openshift-controller-manager-master-e2e-aws/1308798128467480576 passed all tests 
> 
> @Miciah you were under the impression that this PR should continue to fail on `[sig-network] Ingress API should support creating Ingress API operations [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]`?

I wasn't sure.  It would depend on whether the problem was that ingress-to-route was adding a status entry or whether it was because ingress-to-route was deleting the test's status entry.  If the test is passing now, it must have been the latter (as that is the behavior that I just changed).  Let's rerun the test to be sure though.

/test e2e-aws

**Comment by sgreene570** [comment]
Test has passed again! :partying_face: 
/lgtm

**Comment by sgreene570** [comment]
@Miciah is the hold on this PR still necessary?

**Comment by Miciah** [comment]
We can cancel the hold.  Thanks!
/hold cancel
